### ISSUE TITLE
Exporter templates may get overridden by previous version during rolling update

### ISSUE THREAD

**Issue by megglos** [camunda/camunda#14863]
**Describe the bug**

If a zeebe update contains changes to the elastic/opensearch exporter templates, these can temporarily get overridden by the previous version during a rolling update if a pod with the previous version restarts in the process.
If this happens and index requests from a pod with the new version are sent to elastic/opensearch, the index request might fail due to a schema mismatch.

**To Reproduce**
Porbably very hard to time 😅
1. Start a rolling update
2. kill a pod that still runs the previous version
3. when it restarts it will override the templates with the previous state
4. Any index request from a pod with the new version will now fail if the last restarted pod was the previous version

**Expected behavior**

Schema templates should either be versioned, so older templates cannot override newer or they should use matchers that include the version of the schema, so they can't conflict with each other and an index is only created if a template with a matching version is present.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
2023-10-25 13:42:13.924 [Broker-1] [zb-fs-workers-2] [Exporter-2] WARN 
      io.camunda.zeebe.broker.exporter.elasticsearch - Unexpected exception occurred on periodically flushing bulk, will retry later.
io.camunda.zeebe.exporter.ElasticsearchExporterException: Failed to flush bulk request: [Failed to flush 89 item(s) of bulk request [type: strict_dynamic_mapping_exception, reason: mapping set to strict, dynamic introduction of [sequence] within [_doc] is not allowed]]
	at io.camunda.zeebe.exporter.ElasticsearchClient.throwCollectedBulkError(ElasticsearchClient.java:183) ~[zeebe-elasticsearch-exporter-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.exporter.ElasticsearchClient.exportBulk(ElasticsearchClient.java:166) ~[zeebe-elasticsearch-exporter-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.exporter.ElasticsearchClient.flush(ElasticsearchClient.java:107) ~[zeebe-elasticsearch-exporter-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.exporter.ElasticsearchExporter.flush(ElasticsearchExporter.java:166) ~[zeebe-elasticsearch-exporter-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.exporter.ElasticsearchExporter.flushAndReschedule(ElasticsearchExporter.java:152) ~[zeebe-elasticsearch-exporter-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) [zeebe-scheduler-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) [zeebe-scheduler-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) [zeebe-scheduler-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:109) [zeebe-scheduler-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) [zeebe-scheduler-8.2.13.jar:8.2.13]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:205) [zeebe-scheduler-8.2.13.jar:8.2.13]
```

</p>
</details>

**Environment:**
- OS: 
- Zeebe Version: Updating from 8.1.x to 8.2.x
- Configuration: Elastic Exporter, Opensearch probably affected as well
- related support ticket https://jira.camunda.com/browse/SUPPORT-18982

**Comment by remcowesterhoud** [comment]
ZPA Triage:
- Impact is mid. Users could notice it, but the problem will correct itself automatically.
- This problem is in both the ES Exporter as the OS Exporter
- It seems we already put the version number inside of the index name. I'm not sure how it could occur that a broker of an older version overrides the template of the broker of a newer version.
- Requires further analysis to determine how this happens
- Marking it as `later` as we have higher impact bugs to worry about.

**Comment by remcowesterhoud** [comment]
Just had another quick look and came across [this slack thread](https://camunda.slack.com/archives/C037W9NMATG/p1698242697063539) 👀 

I guess this is not about the index name, but about the index pattern that's defined in our templates. These patterns match on `zeebe-record_valuetype_*`. The indexes follow naming pattern `zeebe-record_valuetype_version_date`. As the version is not part of the pattern it could occur that exported record match to the index template of a different version.

**Comment by rodrigo-lourenco-lopes** [comment]
This issue has happened again in one of our internal cluster. The same pattern has happened with the templates having been overwritten by a previous version. [#inc-inc-timer-not-getting-triggered](https://camunda.slack.com/archives/C07RY17GX24)

**Comment by nicpuppa** [comment]
This issue happened again on a prod cluster. [#inc-exporting-is-lagging-behind]
(https://app.slack.com/client/T0PM0P1SA/C07UFER71H9)

**Comment by EuroLew** [comment]
https://github.com/camunda/camunda/pull/24483 should fix this issue

**Comment by koevskinikola** [comment]
Hey @megglos @remcowesterhoud @EuroLew, is there a workaround for this issue?

Seems like this happened during an [upgrade from 8.5 to 8.6 on SaaS](https://camunda.slack.com/archives/C091ZFQSYQ0). From what I can see, the fix was only provided for 8.7.

UPDATE: I found a workaround [here](https://camunda.slack.com/archives/C07UFER71H9/p1730446884143389) (internal link).

TL;DR:

We need to apply the missing properties to the affected indices manually. The code is the following:

```sh
curl -u $ES_USERNAME:$ES_PASSWORD -XPUT 'localhost:9200/zeebe-record*$ZEEBE_VERSION*/_mapping' -H "Content-Type: application/json" -d '{
  "properties": {
         "operationReference": {
          "type": "long"
        }
  }
}'
```

**Comment by kristinkomschow** [comment]
@megglos @remcowesterhoud @EuroLew
Is this something we could backport to 8.6? As @koevskinikola mentioned, this was encountered in an 8.5 to 8.6 upgrade and there are worries this will happen in production as well.
